data-structures/hashmap.md

```diff
@@ -295,3 +298,55 @@ The impact on the data structure layout imposed by the use of buckets in the IPL
 ### Security
 
 As yet, there is no known hash collision attack vector against IPLD data structures. There may conceivably be use cases where user-input is able to impact `key`s and collisions against a chosen `hashAlg` are practical. In such cases, an IPLD HashMap could be built whereby it reaches its maximum depth of `(digestLength x 8) / bitWidth` quickly and further colliding additions cause errors and possible denial of service. The current use-cases of IPLD do not lend themselves to denial of service attacks of this kind. Further practical application and research may change this understanding and dictate the need for hash algorithms with large byte output and/or cryptographic hash algorithms without known collision flaws.
+
+## Appendix: Filecoin HAMT Variant
+
+The [Filecoin Project blockchain](https://filecoin-project.github.io/specs/) makes use of a HAMT that uses the same HAMT with buckets and CHAMP mutation semantics as outlined in this document. It encodes directly as [DAG-CBOR](../block-layer/codecs/dag-cbor.md) but uses a different block layout to the one specified here. This section documents the specific ways that the Filecoin HAMT variant differs from this specification. IPLD HashMap implementations may be able to implement a form that provides compatibility with Filecoin when requested by the user.
+
+The reference Go implementation for the Filecoin HAMT is used by the [Lotus](https://lotu.sh/) client and is available at <https://github.com/ipfs/go-hamt-ipld>.
+
+### Implicit and fixed parameters
+
+The Filecoin HAMT _does not_ use an explicit root block (`HashMapRoot`) to encode its parameters within the data. Instead it is expected that consumers of the data understand the parameters from a combination of the Filecoin specification and versioning of the blockchain over time. All HAMT nodes take the same form, there is no differentiation for a root node and an implementation must bring implicit parameters when decoding each node.
+
+* `hashAlg`: The hash algorithm used by the Filecoin HAMT is the x64 form of the 64-bit [MurmurHash3](https://github.com/aappleby/smhasher). _(TODO: this may be changed to SHA2-256 prior to Filecoin mainnet: https://github.com/filecoin-project/specs-actors/issues/517)_
```

Comment:
The Rust implementation seems to use the 128-bit variant of the x64 Murmur3 hash: https://github.com/ChainSafe/forest/blob/b9ae7e8f53aeecaf2f3a7546dcdf40bb8b8ff55b/ipld/hamt/src/node.rs#L17

Reply:
uh oh https://github.com/ipfs/go-hamt-ipld/blob/21886a1edb600df1cc6cead007fcba408f47af91/hash.go#L53

I'm pretty sure the 64 is _not_ a subset of the 128, so if they're both implemented correctly that HAMT shouldn't be able to read the nodes written by the Go HAMT. But, murmur3 is notorious for cross-impl compatibility problems!

Reply:
I'll talk with the Forest folks about that.

Reply:
I'm pretty confident it is, we have compatibility tested against go impl for this. There is no 64 bit hash value btw as far as I can see. Maybe you are mixing up the 64 bit target with the hash bit length?

(I will look a bit more into to confirm though)

Reply:
heh, yeah I just had my head in there too. There's this comment just above what you pasted:

```
// digest64 is half a digest128.
type digest64 digest128
```

and the reference impl only has a 128 for x64 anyway so I guess it's a subset! https://github.com/aappleby/smhasher/blob/master/src/MurmurHash3.cpp :thumbsup:

Reply:
This was something I didn't understand though when I was working on it, what would be the reason for not keeping all 128 bytes? Why intentionally limit the max depth? (I did overlook that this could be a potential inconsistency because the go implementation will hit max depth sooner and there doesn't seem to be a negative with keeping around more hashed bytes as they are never persisted)

Reply:
No good reason I think, you have to have a pretty large amount of data to hit max depth though (unless you can force collisions as is being discussed in that SHA2-256 proposal thread). For 64 bits you get a maximum of 12 levels if you're taking 5 at each level (as Filecoin is), with 32 elements at each level, that's 32^12, then 3 at the leaves, which is `55340232221128790000` for a maximally full tree assuming a completely random distribution of hashes .. I think. So we're talking very large numbers with good randomness so maybe it was just a case of :shrug:, it's big enough for now (famous last words).